apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerHealthChecker.java

```diff
@@ -24,36 +24,49 @@
  */
 package co.elastic.apm.agent.report;
 
-import co.elastic.apm.agent.context.LifecycleListener;
-import co.elastic.apm.agent.impl.ElasticApmTracer;
 import co.elastic.apm.agent.util.ExecutorUtils;
+import co.elastic.apm.agent.util.Version;
+import com.dslplatform.json.DslJson;
+import com.dslplatform.json.JsonReader;
+import com.dslplatform.json.MapConverter;
+import com.dslplatform.json.Nullable;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.net.HttpURLConnection;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Future;
 import java.util.concurrent.ThreadPoolExecutor;
 
-public class ApmServerHealthChecker implements Runnable, LifecycleListener {
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+public class ApmServerHealthChecker implements Callable<Version> {
     private static final Logger logger = LoggerFactory.getLogger(ApmServerHealthChecker.class);
 
     private final ApmServerClient apmServerClient;
+    private final DslJson<Object> dslJson = new DslJson<>();
 
     public ApmServerHealthChecker(ApmServerClient apmServerClient) {
         this.apmServerClient = apmServerClient;
     }
 
-    @Override
-    public void start(ElasticApmTracer tracer) {
+    public Future<Version> checkHealthAndGetMinVersion() {
         ThreadPoolExecutor pool = ExecutorUtils.createSingleThreadDeamonPool("apm-server-healthcheck", 1);
-        pool.execute(this);
-        pool.shutdown();
+        try {
+            return pool.submit(this);
+        } finally {
+            pool.shutdown();
+        }
     }
 
+    @Nullable
     @Override
-    public void run() {
-        apmServerClient.executeForAllUrls("/", new ApmServerClient.ConnectionHandler<Void>() {
+    public Version call() {
+        List<Version> versions = apmServerClient.executeForAllUrls("/", new ApmServerClient.ConnectionHandler<Version>() {
             @Override
-            public Void withConnection(HttpURLConnection connection) {
+            public Version withConnection(HttpURLConnection connection) {
```

Comment:
What do you say about always returning a version from this method (eg version 0.0.0 on failures), so always having a list of versions at the size of configured server URLs? Then, we don't have to limit ourselves to look only at the minimal version, we can use the `errorCount` state to always know **the version of the currently connected server**.
We only need to make sure we use shuffle the URL list once and use the same shuffle here and for the other users.

Reply:
Forget about it- not THAT required: if we have a bunch of APM servers with different versions (shouldn't be a typical case unless during an APM server cluster upgrade), we would probably want our agents to behave the same with all of them, so the minimal version is actually probably what we are looking for